### ISSUE TITLE
Practical steps towards Shapely 2.0

### ISSUE THREAD

**Issue by jorisvandenbossche** [shapely/shapely#962]
TODO before final release:

- [ ] update list of contributors again

TODO items for the Shapely / PyGEOS merge:

* [x] Merge packages into a single shapely package:
  * [x] Merge the python source code of both packages into a single, minimal working `shapely` package -> https://github.com/Toblerity/Shapely/pull/1237
  * [x] Consolidate both `strtree.py` files -> https://github.com/Toblerity/Shapely/pull/1251
* [x] Continuous Integration:
  * [x] Update and consolidate the github action workflow for testing -> https://github.com/Toblerity/Shapely/pull/1241
  * [x] Update the github action workflow for wheel building / releasing -> https://github.com/Toblerity/Shapely/pull/1253
  * [x] Ensure Travis CI is running (after moving to the shapely org?)
* [x] Packaging:
  * [x] Consolidate the `setup.py` parameters -> https://github.com/shapely/shapely/pull/1376
  * [x] Consolidate pyproject.toml and setup.cfg
  * [x] Consolidate MANIFEST.in, ensure sdist/wheel creation is working -> https://github.com/Toblerity/Shapely/pull/1249
  * [x] Ensure versioneer is working and has a correct shapely 2.0.0.dev+.. version 
* [ ] Tests
  * [ ] Move `/tests` to `/shapely/tests` -> [#1257](https://github.com/shapely/shapely/pull/1257), [#1275](https://github.com/shapely/shapely/pull/1275), ...
  * [ ] Combine / deduplicate tests where useful
* [ ] Documentation
  * [ ] Update docstrings
  * [x] Merge the sphinx documentation sources -> https://github.com/shapely/shapely/pull/1377, https://github.com/shapely/shapely/pull/1402

---

Opening an issue to discuss the more practical aspects of the work on Shapely 2.0 (https://github.com/shapely/shapely-rfc/pull/1/). Changes needed in Shapely 1.8 (deprecations, etc) related to this are tracked in https://github.com/Toblerity/Shapely/issues/932

The general route I am now trying out is to, iteratively, refactor the shapely code in-place. But @caspervdw mentioned at https://github.com/Toblerity/Shapely/issues/782/#issuecomment-592058242 also the option of starting "fresh" (for the code, not for the existing tests and things like that).

For both ways, the work can be done in a separate branch, and then using the normal PR workflow (with PRs targetting that branch). This branch has already been created: https://github.com/Toblerity/Shapely/tree/shapely-2.0 (but is already a bit behind master). 
**First question**: how to keep this branch updated while master is still for 1.8? I would like to propose to *initially* rebase `shapely-2.0` regularly on master. In an initial phase where there is still a lot of activity for 1.8 on master and only few people contribute to the `shapely-2.0` branch, this seems easier / cleaner (there are eg also still some things to be merged in master on which we depend for shapely 2.0). Once the development focus shifts more to shapely-2.0, we can use the normal merging strategy already being used for the different branches right now. Does that sound good?
(I can volunteer to do the rebases / fix conflicts :-))

At some point we should probably move the source code of PyGEOS into Shapely (I would propose to do this in its entirety, and keeping its commit history (for proper attribution), and then clean-up / integrate in subsequent steps). 
But initially, I am actually already testing some things (eg the Geometry subclasses) with using pygeos as an external package.

cc @sgillies @mwtoews @tomplex @caspervdw

---

Practical steps (to further fill in while we go, but to keep track of the items we already have/know):

* [x] Remove the functionality deprecated in Shapely 1.8 (https://github.com/Toblerity/Shapely/issues/932)
  * [x] Mutability -> PR https://github.com/Toblerity/Shapely/pull/960
  * [x] Adapter/proxy classes -> PR https://github.com/Toblerity/Shapely/pull/961
  * [x] Ctypes / array interface
* [x] Refactor the Geometry classes as subclasses of pygeos.Geometry -> https://github.com/Toblerity/Shapely/pull/983
* [x] Remove `shapely.geos.lgeos` and its usage (use pygeos where needed instead) -> https://github.com/Toblerity/Shapely/pull/1163
* [x] Move all above changes into the `main` branch (-> https://github.com/Toblerity/Shapely/pull/1211)
* [x] Move the source code of PyGEOS into Shapely (-> https://github.com/Toblerity/Shapely/pull/1221)
* [ ] ...

**Comment by jorisvandenbossche** [comment]
To give an update on this: the geometry classes already have been refactored to subclass pygeos.Geometry a while ago, and in the meantime I have been refactoring most of the other parts of shapely as well to remove any direct use of the ctypes based `lgeos` and `impl`. 

The two main remaining places where it is used are:

- `shapely.ops` -> I have a PR to refactor this (https://github.com/Toblerity/Shapely/pull/1065), but this still needs some additions to pygeos as well
- `STRtree` -> this can eventually be replaced with the pygeos STRtree implementation, but we first need to decide on the exact API regarding the return values. I started a PR with a concrete proposal (for shapely 1.8, and once agreed can be applied to shapely-2.0 branch as well): https://github.com/Toblerity/Shapely/pull/1064

A few smaller other cases as well (like the validation module, pickling linearrings), will look into fixing those as well.

So in summary, I think we are close (once the above are done) to be able to remove full of `shapely.geos.lgeos` and all ctypes usage (at which point Shapely is basically a pure python package fully depending on pygeos for all GEOS interaction). 

---

With that said, I would like to bring up the next step in the outline above: **"Move the source code of PyGEOS into Shapely"**

I think it would be a good time to start consolidating development (and discussions, interactions, etc) in a single repo (eg after a next release of pygeos).  

Practically speaking, I was thinking about:

- Move the full pygeos git repository into the shapely repository (so we can preserve history for the pygeos files, and commit author recognition, etc). I need to check a bit how this would exactly work, but I think with eg git-subtree this should be possible.
- The first step would simply add the pygeos source, so both are side by side in the same repo. Then we can progressively move things to the correct place in the consolidated repo.
  - The exact file organization is something we should then still discuss. There are also some aspects where both repos use different "practices" (like pyproject.toml, versioneer, ..) that need to be discussed/consolidated
- Regarding issues: it might be useful to migrate open issues from pygeos to Shapely, at some point. We can either do this manually, or try to find an automated way (github's issue migration is only possible for repo's in the same org).

Thoughts about this?

**Comment by jorisvandenbossche** [comment]
I had a chat with @caspervdw about this last week, and trying to summarize a few thoughts:

- Move Shapely 1.8 forward (https://github.com/Toblerity/Shapely/issues/1066) so a beta/final version can be released and its development can be moved to a `maint-1.8` branch
- At that point, the master branch can become the current `shapely-2.0` branch, and then we can move the pygeos source code and development into Shapely

Basically we were thinking that it's probably important to only move pygeos here when the development on it can happen in the master branch, to ease the development workflow. 
(because there will still be quite some work needed to integrate both code bases, and while doing that we will also still further develop the pygeos C code, and this will all be easier to work on if it's on the master branch, I think)

**Comment by jorisvandenbossche** [comment]
Now Shapely 1.8 has been released, and a `maint-1.8` branch has been created, I would like to move forward with the next steps outlined in this issue.

> At that point, the master branch can become the current `shapely-2.0` branch, and then we can move the pygeos source code and development into Shapely

For the step of moving the pygeos source code into the Shapely repo, I would propose the following workflow:

1. Move the full pygeos git repository into the shapely repository as a `/pygeos` subdir using git-subtree
    * This way we can can preserve history for the pygeos files, and commit author recognition, etc
    * Initially adding this as a subdirectory should avoid file name clashes of top-level files in this first step (things like `setup.py` etc)
2. This first step simply add the pygeos source, so both are side by side in the same repo. Then we can progressively move things to the correct place in the consolidated repo, both for the actual package code as for development workflow related files (CI etc). Different aspects, and each of those will require some discussion:
    1. Merge the actual python and C source files (to discuss: package layout, etc)
    2. Merge the packaging and development setup (setup.py, versioning, CI, etc)
    3. Merge the tests
    4. Merge the docs
 
Those different items of step 2) will require some more discussion. I will open a separate issue for the package layout, and discuss some of the development workflow issues here (see comment below). 

I would like to have done the minimal basics for 2.i and 2.ii as fast as possible, so we have the main branch in an installable status as fast as possile (and can start doing shapely 2.0.0-alpha releases as quick as possible to test things out). 

Initially we also will still want to do some updates in pygeos or make new releases of it (since many people are already using it), but I would propose that once step 1 and 2.ii is done, all substantial changes are first committed to shapely, and then backported to pygeos when needed/useful. @caspervdw would that work for you?

**Comment by jorisvandenbossche** [comment]
Some further discussion points for step 2.ii regarding merging the packaging and development setup.

**Consolidation packaging setup and development workflows**


- CI: 
  - pygeos has moved Windows tests from Appveyor to github actions as well. I would propose to follow that lead (one CI service less to worry about)
  - pygeos actually still has some Travis CI setup for more exotic platforms (ppc64le, s390x)
- Linting
    - in pygeos we use flake8, black and isort in a pre-commit config file. Are people OK with using those tools here as well?
- Versioning:
    - Shapely currently uses a simple approach of a harcoded version number in `__init__` that is updated manually on releases. In PyGEOS we use `versioneeer` (https://github.com/python-versioneer/python-versioneer) which uses the git tag to automatically determine the version number. Personally, I like the approach of using git tags (it's also being used in geopandas, pandas, etc): when doing a release you just add a tag (which you need to do anyway) and the versioning follows from that + you also get a development version (like pandas version `1.4.0.dev0+973.g82f5395e0c`). Another alternative for this approach would be `setuptools_scm` (https://github.com/pypa/setuptools_scm)
- Wheel building
    - cibuildwheel: we made some effort setting this up in PyGEOS (see https://github.com/pygeos/pygeos/blob/9dcb9969f028eea90e34238b2746272105a7ff5e/.github/workflows/release.yml#L39-L134). I would propose to keep using this in the merged repo instead of the separate shapely-wheels repo. Cibuildwheel has some advantages: well maintained, easier to use, can be included in the main repo (so you can more easily test a PR that might impact wheel building; no need to synchronize the shapely repo inside shapely-wheels), and some other packages I know also started or are starting to use it (eg scikit-learn, numpy).
- Release / publishing:
  - In addition to building wheels, in pygeos we also automated the building of sdist / uploading to github release / to PyPI for releases (see https://github.com/pygeos/pygeos/blob/9dcb9969f028eea90e34238b2746272105a7ff5e/.github/workflows/release.yml#L136-L181). We have been using a similar approach in geopandas as well, and it makes releasing a very smooth experience.

**Comment by jorisvandenbossche** [comment]
Heads up: the first step (merging the pygeos source code and history into this repo) has been done!

This will require lots of follow-up to clean up and consolidate the merged repo (will make a more detailed checklist in the top post), but I started with a first step to get the package working again: https://github.com/Toblerity/Shapely/pull/1237

**Comment by jorisvandenbossche** [comment]
@sgillies I tagged the main branch with 2.0.0.dev (the commit that merged in pygeos, since there was no good commit directly after 1.8 to tag, shouldn't also really matter), so with the current versioneer setup, you get this version on the main branch now:

```
In [1]: import shapely

In [2]: shapely.__version__
Out[2]: '2.0.0.dev0+4.g0d3d811'
```

**Comment by jorisvandenbossche** [comment]
Now that the STRtree API changes are mostly settled for 2.0 (https://github.com/shapely/shapely/issues/1386), I think the main open issue/PR is https://github.com/shapely/shapely/pull/1414 (about the `Geometry()` class constructor).

Next to that we also listed the credits update and changelog as requirements for a first 2.0 alpha release. I have started https://github.com/shapely/shapely/pull/1442 for the changelog.

I would propose to do cut a first alpha 0 after merging https://github.com/shapely/shapely/pull/1414? Having a first alpha would allow to test the release / wheel building workflow, update the conda recipes (and make it easier to use it in CI for example in geopandas). And we can still only communicate about it once the changelog is in a better state.

**Comment by jorisvandenbossche** [comment]
@sgillies @caspervdw @brendan-ward @mwtoews as mentioned above (and discussed before), I am planning to tag a first alpha release after merging https://github.com/shapely/shapely/pull/1414 (which should be ready to be merged). Please speak up if there are any concerns about this!

**Comment by brendan-ward** [comment]
I think it would be good to have #1451 and #1391 in first.  Do we also want to make sure that one of #1426 / #1430 is in so that `setup.py` and related are a bit more stable before the alpha?

**Comment by jorisvandenbossche** [comment]
> Do we also want to make sure that one of #1426 / #1430 is in so that `setup.py` and related are a bit more stable before the alpha?

In _theory_ that should only be moving the same information around to another file / config format. Of course in practice there might be small nuances. I think the main problem is that pyproject.toml (https://github.com/shapely/shapely/pull/1426) is blocked by dropping Python 3.6 support. 
But personally I don't see those as critical for an alpha release.

**Comment by mwtoews** [comment]
I'm investigating the static project metadata PRs, with a preference with [PEP 621](https://peps.python.org/pep-0621/) as a final place for the "boring parts".

**Comment by jorisvandenbossche** [comment]
All of https://github.com/shapely/shapely/pull/1451, https://github.com/shapely/shapely/pull/1391 and https://github.com/shapely/shapely/pull/1426 are merged (as well as the update to require Python >= 3.7), so I am planning to tag a first alpha.

**Comment by jorisvandenbossche** [comment]
The automated release / packaging seems to have worked for the github relaese: https://github.com/shapely/shapely/releases/tag/2.0a1

However:

- The upload from github actions to PyPI failed(https://github.com/shapely/shapely/runs/7648698789?check_suite_focus=true#step:7:12), although the PYPI_API_TOKEN has been set up. There is a warning about a missing "pypi-" prefix. I might have missed that, will try to update it.
- The upload to PyPI from travis (linux aarch64 wheels) is also failing simply because we forgot to add the encrypted password for this. I am looking creating an encrypted token, however I am running into an issue that seems to happen for old repos: https://travis-ci.community/t/travis-encrypt-data-too-large-for-pypi-tokens-with-older-repos/5792

**Comment by jorisvandenbossche** [comment]
Update: for github PyPI upload, it was just a small mistake: the secret was named differently than what we had in the workflow file .. I updated that, and re-ran the workflow, and now it is working: https://pypi.org/project/shapely/2.0a1/

**Comment by sgillies** [comment]
@jorisvandenbossche have you announced on Twitter yet?

**Comment by jorisvandenbossche** [comment]
I first wanted to get the conda-forge packages done (but feel free to already announce, that's not critical)

**Comment by soxofaan** [comment]
To whom it might concern: having a dependency on `geopyspark` (or just installing it) might blindly drag in shapely 2.0 alpha/beta releases in your environment. See https://github.com/locationtech-labs/geopyspark/issues/712

**Comment by jorisvandenbossche** [comment]
So the conda packages are also ready. If you want to install shapely 2.0 alpha 1, you can do:

```
# using pip
pip install --pre shapely==2.0a1
# using conda /mamba
conda install -c conda-forge/label/shapely_dev shapely
```

Draft release notes can already be seen at https://github.com/shapely/shapely/pull/1442/files#diff-f271474562bb2f8c3bdcce3839a98b1ddf7d918b1a4711c6e04f1b572725f12b, although the existing migration guide for updating to 1.8 / 2.0 might be good to point to as well: https://shapely.readthedocs.io/en/stable/migration.html

**Comment by sgillies** [comment]
@soxofaan thanks for pointing that out!

**Comment by jorisvandenbossche** [comment]
I would like to move forward with a next beta release. 

I think most of the issues we have been discussing have been merged (a couple of renames (`offset_curve`/`parallel_offset`, `quad_segs` keyword, `unary_union` alias, `max_segment_length` keyword, documenting the changed behaviour of GEOS for `offset_curve`, etc). The remaining item is:

* Behaviour of reducers like `unary_union` (`union_all`) for empty or all-None arrays: https://github.com/shapely/shapely/issues/1332. 
  * There is a lot of discussion about this in [#1501](https://github.com/shapely/shapely/pull/1501), and a fix in [#1540](https://github.com/shapely/shapely/pull/1540). 
  * Given this is a PR that might not be merged in the next few days, and given that for shapely users, it is mostly the `unary_union` behaviour that matters, I also opened https://github.com/shapely/shapely/pull/1553 targeted just at `unary_union` to potentially include in 2.0 beta1

In addition, I think those two following items would also benefit from getting a decision:

* https://github.com/shapely/shapely/pull/1231 (do we want to normalize by default or not?)
* https://github.com/shapely/shapely/pull/1522

**Comment by jorisvandenbossche** [comment]
Unless I hear something, I am planning to do a beta release shortly (with https://github.com/shapely/shapely/pull/1553 included)

**Comment by mwtoews** [comment]
One little undiscussed feature that I'm nearly done is to implement a geometry [`__format__` specification](https://docs.python.org/3/library/string.html#formatspec) (e.g. `f"{Point(1, 2):.2f}"` to get `POINT (1.00 2.00)`). While not a crucial feature, it would be nice to put into 2.0. Hopefully I can complete it in another day for review. A draft for implementing https://github.com/shapely/shapely/discussions/1490 is still weeks away, however, so probably won't make it.

**Comment by jorisvandenbossche** [comment]
Practical note: personally I would say that it is still fine to add _new_ features until a first RC release. Because I mostly want to get the beta release out to have more people test Shapely 2.0 to report compatibility / updating issues for code now runs on Shapely < 2, and in theory completely new features shouldn't impact that. 
(so to say: I would be fine with still merging a `__format__` PR after the beta)

**Comment by jjyyxx** [comment]
Would #1443 be a part of the beta? Or would it be included after the beta? Part of my project extends shapely with C code and relies on casting `_geom` as a C pointer. So I need such a C API for migration.

**Comment by jorisvandenbossche** [comment]
@jjyyxx can you give a bit more details about what you are trying to do? (but maybe in #1443 to keep the discussion about a C API there) The `_geom` attribute is accessible, so you should still be able to access that in C and cast it to a C pointer, the same way as you could do with shapely <= 1.8.

**Comment by jorisvandenbossche** [comment]
I tagged and released a beta 1! -> https://pypi.org/project/shapely/2.0b1 (https://github.com/shapely/shapely/releases/tag/2.0b1)

This can be installed with pip or conda:

```
# using pip
pip install --pre shapely==2.0b1
# using conda /mamba
conda install -c conda-forge/label/shapely_dev shapely
```

**Comment by jorisvandenbossche** [comment]
For getting a release candidate and final release for Shapely 2.0, @caspervdw and I were thinking to just put out a target date we can follow (except if new issues come up). 

So one proposal would be: 4 weeks for the beta + 2 weeks for a release candidate, and then a final release.

Given that the first beta was released on Monday October 17, the target dates could then be Monday November 14 for a first release candidate, and if all goes fine Monday November 28 for a final 2.0.0 release.

Thoughts on such a timeline?

We can still merge entirely new features until the first RC (eg https://github.com/shapely/shapely/pull/1556, https://github.com/shapely/shapely/pull/1548), and we can still do API changes (if needed) before the RC (and then ideally still release a next beta to include those in a release before the RC)

**Comment by sgillies** [comment]
@jorisvandenbossche sounds great!

**Comment by jorisvandenbossche** [comment]
Released a beta 2 with some of the latest fixes: https://github.com/shapely/shapely/releases/tag/2.0b2

**Comment by jorisvandenbossche** [comment]
The date for a first RC passed, and while I am still planning to have that out this week, there are a few PRs that are currently milestoned for 2.0 (https://github.com/shapely/shapely/pulls?q=is%3Aopen+is%3Apr+milestone%3A2.0) that I would still like to include (or at least make some decision on whether to include it for 2.0 or not):

* https://github.com/shapely/shapely/pull/1604 
* https://github.com/shapely/shapely/pull/1559 
* https://github.com/shapely/shapely/pull/1548

**Comment by idanmiara** [comment]
Hi all,
Thanks for making shapely even greater!

Did you consider adding any kind of a srs/crs attribute to shapely 2.0 objects?

**Comment by jorisvandenbossche** [comment]
> may I propose a 2.0b3? People would love to see concave hulls in action. 

@sgillies You can also see the concave hull in action with a 2.0rc1! ;)

Joking aside, the new `concave_hull` (is indeed awesome) is a slim wrapper around a GEOS function, just as we have many others. So testing this is mostly testing the new GEOS implementation, which is useful of course, but something that will typically require changes in GEOS, not Shapely. So I think in _general_, IMO we shouldn't expect to always do a cycle of test releases whenever we have a release that exposes something new from GEOS. 
(there might be other reasons to still do a beta release at this point, though, I just don't think concave_hull is one)

**Comment by jorisvandenbossche** [comment]
> Did you consider adding any kind of a srs/crs attribute to shapely 2.0 objects?

@idanmiara that is currently considered out of scope for Shapely, and in any case a bigger discussion that we are not going to have for the upcoming 2.0 release (since we are finalizing this). If you would like to bring this up and discuss further, let's do that in a separate issue (maybe first have a look through older issues if there is already something about, I would have to look myself as well)

**Comment by sgillies** [comment]
@idanmiara shapely is (mostly) only a planar geometry library. Separation of concerns is good for software. Coordinate reference systems, features, layers, cartography, all that has been out of scope and I'm convinced that those concerns will always be better left to, for example, GeoPandas.

**Comment by idanmiara** [comment]
@sgillies @jorisvandenbossche

Thanks for you kind replies. 

Before I open a separate issue to discuss this further, I just wanted to know if you see any technical difficulties (API/ABI compatibility for instance) with adding an immutable srs/crs attribute in shapely 2.1 geometry objects?

**Comment by jorisvandenbossche** [comment]
Sorry to be strict, but answering that is for that separate issue ;) (I really want to keep this focused on getting out 2.0). 
And I would certainly encourage you to actually open a separate issue (or maybe discussion: https://github.com/shapely/shapely/discussions). Even if it is for us to just repeat that this is out of scope (but we can go a bit in more detail), it is a useful to have a reference for this question and answer (since you are not the only one wondering this).

**Comment by jorisvandenbossche** [comment]
As a heads up, pending no further comments or feedback, I am planning to cut a first RC tomorrow or on Saturday

**Comment by sgillies** [comment]
@jorisvandenbossche do it!

**Comment by idanmiara** [comment]
RC1 is on PyPi, Thanks @jorisvandenbossche 
https://pypi.org/project/shapely/2.0rc1/

**Comment by jorisvandenbossche** [comment]
Yes, and in the meantime everything is packaged as well. So you can install and test the rc1 with:

```
# using pip
pip install --pre shapely==2.0rc1
# using conda /mamba
conda install -c conda-forge/label/shapely_dev shapely
```

**Comment by jorisvandenbossche** [comment]
Did a quick RC2 to test the latest changes to packaging (the update to pyproject.toml, building aarch64 wheels on CircleCI).

**Comment by felixdivo** [comment]
RC2 seems to work fine in our project. 🥳

**Comment by jorisvandenbossche** [comment]
General note: the final 2.0.0 release is planned for tomorrow (2 weeks after the RC)

@felixdivo thanks a lot for testing!

**Comment by idanmiara** [comment]
Hi all,
I was wondering what is the estimated timeline for `2.0.1` and which opened PRs are considered for that version.
Thanks!

**Comment by jorisvandenbossche** [comment]
No concrete timeline, but I hope to do a push this weekend, and we should probably release 2.0.1 shortly with whatever we already have (I think only https://github.com/shapely/shapely/issues/1707 is a remaining reported regression that we should certainly ensure to include)

**Comment by idanmiara** [comment]
I think that https://github.com/shapely/shapely/issues/1711 is also a bug/regression that should be addressed in `2.0.1` (as only shapely 2 uses relative imports)

**Comment by jorisvandenbossche** [comment]
I am planning to do a first 2.0.1 release with the fixes we have merged at the moment (https://github.com/shapely/shapely/pull/1737). There are a few more fixes in the pipeline, but we can do a 2.0.2 not much after. 

> I think that #1711 is also a bug/regression that should be addressed in `2.0.1` (as only shapely 2 uses relative imports)

@idanmiara I think that's only an issue for contributors, not for end users? (so I wouldn't consider that as critical for 2.0.1)

**Comment by jorisvandenbossche** [comment]
Also going to close this, since 2.0.0 is released now. The remaining follow-up issues listed in the top post (tests, docs) have its own issue now. Will open a separate issue for coming bug-fix releases.

**Comment by idanmiara** [comment]
> > I think that #1711 is also a bug/regression that should be addressed in `2.0.1` (as only shapely 2 uses relative imports)
> 
> @idanmiara I think that's only an issue for contributors, not for end users? (so I wouldn't consider that as critical for 2.0.1)

As a user I would expect it to be easy to run/debug doctests to help me understand how a function works, thus I would prioritize it.
